### PR TITLE
[Foxy backport] wait for message (#1705)

### PR DESCRIPTION
**Pull request by Karsten1987**
Backport of #1705

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
Do we have backport for galactic?

File: rclcpp/include/rclcpp/wait_for_message.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ */
+template<class MsgT, class Rep = int64_t, class Period = std::milli>
+bool wait_for_message(
+  MsgT & out,
+  std::shared_ptr<rclcpp::Subscription<MsgT>> subscription,
+  std::shared_ptr<rclcpp::Context> context,
+  std::chrono::duration<Rep, Period> time_to_wait = std::chrono::duration<Rep, Period>(-1))
+{
+  auto gc = std::make_shared<rclcpp::GuardCondition>(context);
+  auto shutdown_callback_handle = context->add_on_shutdown_callback(
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply rclcpp/include/rclcpp/wait_for_message.hpp:47]: bummer. This seems like a potential show stopper for a Foxy backport.

**Comment by Karsten1987** [comment]
My intention here was to see whether we could use the backport to Foxy within ros-control. We're still targeting Foxy at this point. I guess I'll only backport it to Galactic then and come up with a somewhat compatible version of `wait_for_message` within ros2_control.

File: rclcpp/include/rclcpp/wait_for_message.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Karsten1987) [reply rclcpp/include/rclcpp/wait_for_message.hpp]: I believe the functionality is already there in Foxy. I seems to me the previously existing API `on_shutdown` (https://github.com/ros2/rclcpp/blob/foxy/rclcpp/include/rclcpp/context.hpp#L199) is still valid, c.f. https://github.com/ros2/rclcpp/pull/1639

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14965/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9705/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12642/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15128/)

File: rclcpp/include/rclcpp/wait_for_message.hpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ */
+template<class MsgT, class Rep = int64_t, class Period = std::milli>
+bool wait_for_message(
+  MsgT & out,
+  std::shared_ptr<rclcpp::Subscription<MsgT>> subscription,
+  std::shared_ptr<rclcpp::Context> context,
+  std::chrono::duration<Rep, Period> time_to_wait = std::chrono::duration<Rep, Period>(-1))
+{
+  auto gc = std::make_shared<rclcpp::GuardCondition>(context);
+  auto shutdown_callback_handle = context->on_shutdown(
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/wait_for_message.hpp:47]: nit: `on_shutdown` returns callback itself but handle.

```suggestion
  auto shutdown_callback = context->on_shutdown(
```

**Review by fujitatomoya — APPROVED** [review]
@Karsten1987 that was quick 😄 thanks!

**Comment by Karsten1987** [comment]
PR job as well as Windows test failures are unrelated to this change.

**Pull request merged**